Add auto support

tugsbayasgalan opened the pull request:
Differential Revision: D76826772

zhxchen17 commented on torch/export/experimental/__init__.py:
```diff
@@ -293,18 +295,35 @@ def _exporter_context(*args, **kwargs):  # type: ignore[no-untyped-def]
-            else:
-                raise RuntimeError(f"Unknown fallback policy: {fallback}")
-            ep = torch.export.export(model, args, kwargs)
+                if len(fallbacks) == 0:
```
you can jus inline the branch here without len() == 0

zhxchen17 commented on torch/export/experimental/__init__.py:
```diff
@@ -113,6 +113,7 @@ def wrapper(*args, **kwargs):
 class _ExportMethod:
     overloads: dict[str, torch.export.ExportedProgram]
     fallbacks: list[torch.export.ExportedProgram]
+    example_inputs: torch.export.AdditionalInputs() = None
```
This is an impl detail and shouldn't be exposed to end user.

zhxchen17 commented on torch/export/experimental/__init__.py:
```diff
@@ -293,18 +295,35 @@ def _exporter_context(*args, **kwargs):  # type: ignore[no-untyped-def]
+                
+                    if not example_inputs.verify(ep):
+                        ep = torch.export.export(model, args, kwargs, dynamic_shapes=example_inputs)
+                        fallbacks[0] = ep 
```
In general overwriting fallbacks seem bad because it's not only shapes that can change across exports. Every time any guard changes, there should be a new fallback produced instead of using the new EP to replace the old EP (which may depend on non-shape invariants).

For example, if I pass 3 inputs in the first run and pass 2 inputs in the second run, we will have very confusing result where we couldn't replay the ExportMethod with the initial 3 inputs anymore. 

At least we should append to the fallback list every time when some metadata on tensor (or the number of tensors) is changed. We should probably never overwrite a fallback list (because non-strict export is unsound and we cannot safely do so anyway)

avikchaudhuri committed:
fix dynamic shapes for kwargs

Differential Revision: D72350333

tugsbayasgalan committed:
Add auto support (#156189)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/156189

Test Plan:
CI

Rollback Plan:

Differential Revision: D76826772

zhxchen17 commented on torch/export/experimental/__init__.py:
```diff
@@ -293,18 +329,51 @@ def _exporter_context(*args, **kwargs):  # type: ignore[no-untyped-def]
-            else:
-                raise RuntimeError(f"Unknown fallback policy: {fallback}")
-            ep = torch.export.export(model, args, kwargs)
+                if len(fallbacks) == 0:
```
```suggestion
                else:
```

titaiwangms commented on test/onnx/exporter/test_small_models_e2e.py:
```diff
@@ -356,7 +356,7 @@ def forward(self, a, b, kw1, kw2):
             # We are specifying dynamism on the first kwarg even though user passed in
             # different order
-            dynamic_shapes=(None, {0: dim}, {0: dim_for_kw1}, None),
+            dynamic_shapes=(None, {0: dim}, None, {0: dim_for_kw1}),
```
Could you reason this?
